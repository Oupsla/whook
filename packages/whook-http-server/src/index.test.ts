import { HTTPRouterService } from '@whook/http-router';
import initHTTPServer from './index';
import axios from 'axios';
import net from 'net';
import YError from 'yerror';

describe('initHTTPServer', () => {
  const PORT = 7777;
  const HOST = 'localhost';
  const log = jest.fn();
  const httpRouter = jest.fn();

  beforeEach(() => {
    log.mockReset();
    httpRouter.mockReset();
  });

  test('should work with keepalive connections', async () => {
    const httpServer = await initHTTPServer({
      HOST,
      PORT,
      log,
      httpRouter: (httpRouter as unknown) as HTTPRouterService,
    });

    await httpServer.dispose();

    expect({
      logCalls: log.mock.calls,
      httpRouterCalls: httpRouter.mock.calls,
    }).toMatchSnapshot();
  });

  test('should work with instantly destroyed connections', async () => {
    const httpServer = await initHTTPServer({
      ENV: {
        DESTROY_SOCKETS: '1',
      },
      HOST,
      PORT,
      httpRouter,
    });

    await httpServer.dispose();

    expect({
      logCalls: log.mock.calls,
      httpRouterCalls: httpRouter.mock.calls,
    }).toMatchSnapshot();
  });

  test('should proxy server fatal errors', async () => {
    const httpServer = await initHTTPServer({
      ENV: {
        DESTROY_SOCKETS: '1',
      },
      MAX_CONNECTIONS: 4,
      HOST,
      PORT,
      httpRouter,
    });

    try {
      httpServer.service.emit('error', new YError('E_ERROR'));

      await httpServer.fatalErrorPromise;
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      await httpServer.dispose();

      expect({
        errorCode: err.code,
        errorParams: err.params,
        logCalls: log.mock.calls,
        httpRouterCalls: httpRouter.mock.calls,
      }).toMatchSnapshot();
    }
  });

  test('should proxy server close errors', async () => {
    const httpServer = await initHTTPServer({
      ENV: {
        DESTROY_SOCKETS: '1',
      },
      MAX_CONNECTIONS: 4,
      HOST,
      PORT,
      httpRouter,
    });

    try {
      httpServer.service.close = ((realClose => async cb => {
        await new Promise((resolve, reject) => {
          realClose(err => (err && (reject(err) as any)) || resolve());
        });

        cb(new YError('E_ERROR'));
      })(
        httpServer.service.close.bind(httpServer.service),
      ) as unknown) as typeof httpServer.service.close;

      await httpServer.dispose();
      throw new YError('E_UNEXPECTED_SUCCESS');
    } catch (err) {
      expect({
        errorCode: err.code,
        errorParams: err.params,
        logCalls: log.mock.calls,
        httpRouterCalls: httpRouter.mock.calls,
      }).toMatchSnapshot();
    }
  });

  test('should close even with opened sockets when using DESTROY_SOCKETS', async () => {
    const httpServer = await initHTTPServer({
      ENV: {
        DESTROY_SOCKETS: '1',
      },
      HOST,
      PORT,
      httpRouter,
    });

    const client = net.createConnection({
      host: HOST,
      port: PORT,
    });

    await new Promise((resolve, reject) => {
      client.on('error', reject);
      client.on('connect', resolve);
      client.write('GET / HTTP/1.1\r\n\r\n');
    });

    await httpServer.dispose();

    expect({
      logCalls: log.mock.calls,
      httpRouterCalls: httpRouter.mock.calls,
    }).toMatchSnapshot();
  });

  test('should serve the router', async () => {
    const httpServer = await initHTTPServer({
      ENV: {
        DESTROY_SOCKETS: '1',
      },
      HOST,
      PORT,
      httpRouter: (((_, res) => {
        res.writeHead(200, {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Methods': 'GET,PUT,POST,DELETE',
          'Access-Control-Allow-Headers': 'Content-Type',
        });
        res.end();
      }) as unknown) as HTTPRouterService,
    });

    const { status } = await axios({
      method: 'get',
      url: `http://${HOST}:${PORT}/`,
    });

    await httpServer.dispose();

    expect({
      status,
      logCalls: log.mock.calls,
    }).toMatchSnapshot();
  });
});
