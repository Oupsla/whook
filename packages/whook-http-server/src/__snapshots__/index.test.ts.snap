// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initHTTPServer should close even with opened sockets when using DESTROY_SOCKETS 1`] = `
Object {
  "httpRouterCalls": Array [],
  "logCalls": Array [],
}
`;

exports[`initHTTPServer should proxy server close errors 1`] = `
Object {
  "errorCode": "E_ERROR",
  "errorParams": Array [],
  "httpRouterCalls": Array [],
  "logCalls": Array [],
}
`;

exports[`initHTTPServer should proxy server fatal errors 1`] = `
Object {
  "errorCode": "E_HTTP_SERVER_ERROR",
  "errorParams": Array [],
  "httpRouterCalls": Array [],
  "logCalls": Array [],
}
`;

exports[`initHTTPServer should serve the router 1`] = `
Object {
  "logCalls": Array [],
  "status": 200,
}
`;

exports[`initHTTPServer should work with instantly destroyed connections 1`] = `
Object {
  "httpRouterCalls": Array [],
  "logCalls": Array [],
}
`;

exports[`initHTTPServer should work with keepalive connections 1`] = `
Object {
  "httpRouterCalls": Array [],
  "logCalls": Array [
    Array [
      "warning",
      "üéôÔ∏è - HTTP Server listening at \\"http://localhost:7777\\".",
    ],
    Array [
      "debug",
      "‚úÖ - Closing HTTP server.",
    ],
    Array [
      "debug",
      "‚úîÔ∏è - HTTP server closed!",
    ],
  ],
}
`;
